README.md
# pokedex

This is the pokedex project.

## Development mode

To start the Figwheel compiler, navigate to the project folder and run the following command in the terminal:

```
lein figwheel
```

Figwheel will automatically push cljs changes to the browser. The server will be available at [http://localhost:3449](http://localhost:3449) once Figwheel starts up.

Figwheel also starts `nREPL` using the value of the `:nrepl-port` in the `:figwheel`
config found in `project.clj`. By default the port is set to `7002`.

The figwheel server can have unexpected behaviors in some situations such as when using
websockets. In this case it's recommended to run a standalone instance of a web server as follows:

```
lein do clean, run
```

The application will now be available at [http://localhost:3000](http://localhost:3000).


### Optional development tools

Start the browser REPL:

```
$ lein repl
```
The Jetty server can be started by running:

```clojure
(start-server)
```
and stopped by running:
```clojure
(stop-server)
```


## Building for release

```
lein do clean, uberjar
```

## Deploying to Heroku

Make sure you have [Git](http://git-scm.com/downloads) and [Heroku toolbelt](https://toolbelt.heroku.com/) installed, then simply follow the steps below.

Optionally, test that your application runs locally with foreman by running.

```
foreman start
```

Now, you can initialize your git repo and commit your application.

```
git init
git add .
git commit -m "init"
```
create your app on Heroku

```
heroku create
```

optionally, create a database for the application

```
heroku addons:add heroku-postgresql
```

The connection settings can be found at your [Heroku dashboard](https://dashboard.heroku.com/apps/) under the add-ons for the app.

deploy the application

```
git push heroku master
```

Your application should now be deployed to Heroku!
For further instructions see the [official documentation](https://devcenter.heroku.com/articles/clojure).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-17
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-17
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/pokedex/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/pokedex/compare/0.1.0...0.1.1


test/pokedex/core_test.clj
(ns pokedex.core-test
  (:require [clojure.test :refer :all]
            [pokedex.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj/pokedex/core.clj
(ns pokedex.core
  (:gen-class)
  (:require [org.httpkit.server :as server]
            [clojure.data.json :refer [read-str write-str]]
            [clj-http.client :as client]
            [compojure.core :refer :all]
            [hiccup.page :refer [include-js include-css html5]]
            [taoensso.sente :as sente]
            [taoensso.sente.server-adapters.http-kit
             :refer (get-sch-adapter)]
            [ring.middleware.anti-forgery
             :refer [wrap-anti-forgery]]
            [ring.middleware.defaults :refer
             [wrap-defaults site-defaults]]
            [compojure.route :as route]))

(defn random-pokemon [region] ;; #{:johto :kanto}
  (let [num (case region
              :johto (+ 151 (rand-int 251))
              :kanto (+ 1 (rand-int 151))
              (+1 (rand-int 897)))
        {name "name"
         {sprite "front_default"
          :as sprites} "sprites"
         :as pokemon}(-> "https://pokeapi.co/api/v2/pokemon/%s"
            (format num)
            client/get
            :body ;;also handle status
            (read-str :keyword-fn keyword)
            )]
    {:name name :sprite-url sprite}))

(def mount-target
  (let  [csrf-token
         (force ring.middleware.anti-forgery/*anti-forgery-token*)]
    [:div#app
     [:div#sente-csrf-token {:data-csrf-token csrf-token}]
     ]))

(defn head []
  [:head
   [:meta {:charset "utf-8"}]
   [:meta {:name "viewport"
           :content "width=device-width, initial-scale=1"}]
   (include-css ;;adds bulma css
    "https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.4/css/bulma.min.css")
   (include-css "/css/site.css")])

(defn loading-page []
  (let [pokemon (random-pokemon :kanto)]
    (html5
     (head)
     [:body {:class "body-container"}
      mount-target
      [:h1 (:name pokemon)]
      [:image
       {:src (:sprite-url pokemon)}]
      (include-js "/js/app.js")
      [:script "pokedex.core.init_BANG_()"]])))

(defn index-handler
  [_request]
  {:status 200
   :headers {"Content-Type" "text/html"}
   :body (loading-page)})

(let [{:keys [ch-recv send-fn connected-uids
              ajax-post-fn ajax-get-or-ws-handshake-fn]}
      (sente/make-channel-socket! (get-sch-adapter) {})]

  (def ring-ajax-post                ajax-post-fn)
  (def ring-ajax-get-or-ws-handshake ajax-get-or-ws-handshake-fn)
  (def ch-chsk                       ch-recv) ; ChannelSocket's receive channel
  (def chsk-send!                    send-fn) ; ChannelSocket's send API fn
  (def connected-uids                connected-uids) ; Watchable, read-only atom
  )

(defroutes app-routes
  (GET "/" [] index-handler)
  (GET "/chsk" req (ring-ajax-get-or-ws-handshake req))
  (POST "/chsk" req (ring-ajax-post req))
  (route/not-found "Welcome to Lost"))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (let [prt (Integer/parseInt (or (System/getenv "PORT") "3000"))]
    (-> #'app-routes
        (wrap-defaults site-defaults)
        wrap-anti-forgery
        ring.middleware.keyword-params/wrap-keyword-params
        ring.middleware.params/wrap-params
        ;; ring.middleware.anti-forgery/wrap-anti-forgery
        ;; ring.middleware.session/wrap-session
        (server/run-server {:port prt}))))


project.clj
(defproject pokedex "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [http-kit "2.6.0-alpha1"]
                 [compojure "1.6.2"]
                 [hiccup "1.0.5"]
                 [clj-http "3.12.3"]
                 [org.clojure/data.json "2.4.0"]
                 ;;add datomic
                 [org.clojure/core.async "1.4.627"]
                 [org.clojure/tools.nrepl   "0.2.13"] ; Optional, for Cider

                 [com.taoensso/timbre       "5.1.2"]
                 [ring "1.9.4"]
                 [ring/ring-defaults        "0.3.3"]
                 [com.taoensso/sente "1.16.2"]
                 [org.clojure/clojurescript "1.10.891"
                  :scope "provided"]
                 [pez/clerk "1.0.0"]
                 [venantius/accountant "0.2.5"
                  :exclusions [org.clojure/tools.reader]]
                 [reagent "1.1.0"]
                 [reagent-utils "0.3.3"]
                 [cljsjs/react "17.0.2-0"]
                 [cljsjs/react-dom "17.0.2-0"]

                 ;;re-frame

                 ]
  :resource-paths ["resources" "target/cljsbuild"]
  :source-paths ["src/clj" "src/cljs"]
  :main ^:skip-aot pokedex.core
  :target-path "target/%s"
  ;;:profiles {:uberjar {:aot :all}}

  :cljsbuild
  {:builds {:min
            {:source-paths ["src/cljs"  "env/prod/cljs"]
             :compiler
             {:output-to        "target/cljsbuild/public/js/app.js"
              :output-dir       "target/cljsbuild/public/js"
              :source-map       "target/cljsbuild/public/js/app.js.map"
              :optimizations :advanced
              :infer-externs true
              :pretty-print  false}}
            :app
            {:source-paths ["src/cljs"  "env/dev/cljs"]
             :figwheel {:on-jsload "pokedex.core/mount-root"}
             :compiler
             {:main "pokedex.dev"
              :asset-path "/js/out"
              :output-to "target/cljsbuild/public/js/app.js"
              :output-dir "target/cljsbuild/public/js/out"
              :source-map true
              :optimizations :none
              :pretty-print  true}}
            }}

  :figwheel
  {:http-server-root "public"
   :server-port 3449
   :nrepl-port 7002
   :nrepl-middleware [cider.piggieback/wrap-cljs-repl
                      ]
   :css-dirs ["resources/public/css"]
   :ring-handler pokedex.core/app-routes}

  :profiles {:dev {:repl-options {:init-ns pokedex.repl}
                   :dependencies [[cider/piggieback "0.5.2"]
                                  [binaryage/devtools "1.0.2"]
                                  [ring/ring-mock "0.4.0"]
                                  [ring/ring-devel "1.9.1"]
                                  [prone "2020-01-17"]
                                  [figwheel-sidecar "0.5.20"]
                                  [nrepl "0.8.3"]
                                  [thheller/shadow-cljs "2.14.3"]
                                  [pjstadig/humane-test-output "0.10.0"]

 ]

                   :source-paths ["env/dev/clj"]
                   :plugins [[lein-figwheel "0.5.20"]
]

                   :injections [(require 'pjstadig.humane-test-output)
                                (pjstadig.humane-test-output/activate!)]

                   :env {:dev true}}

             :shadow-cljs {:dependencies [[com.google.javascript/closure-compiler-unshaded "v20210505"]]}

             :uberjar {:hooks [minify-assets.plugin/hooks]
                       :source-paths ["env/prod/clj"]
                       :prep-tasks ["compile" ["cljsbuild" "once" "min"]]
                       :env {:production true}
                       :aot :all
                       :omit-source true}}

  )
